rust/plugin/src/diff/differ.rs

```diff
@@ -1,126 +1,317 @@
+                .get_files_at(heads, Some(&HashSet::from_iter(vec![path.clone()])));
+            for file in files.into_iter() {
+                if file.0 == *path {
+                    ret = Some(file.1);
+                    break;
+                } else {
+                    panic!(
+                        "Returned a file that didn't match the path!?!??!?!?!?!?!?!!? {:?} != {:?}",
+                        file.0, path
+                    );
+                }
+            }
+        }
+        ret
+    }
+
+    /// Loads an ExtResource given a path, using a cache.
+    pub(super) fn load_ext_resource(
+        &self,
+        path: &String,
+        heads: &Vec<ChangeHash>,
+    ) -> Option<Variant> {
+        if let Some(resource) = self.loaded_ext_resources.borrow().get(path) {
+            return Some(resource.clone());
+        }
+
+        let resource_content = self.get_file_at(path, Some(heads));
+        let Some(resource_content) = resource_content else {
+            return None;
+        };
+
+        let Some(resource) = self.get_resource_at(path, &resource_content, heads) else {
+            return None;
+        };
+
+        self
+            .loaded_ext_resources
+            .borrow_mut()
+            .insert(path.clone(), resource.clone());
+		Some(resource)
+    }
+
+	/// Computes the diff between the two sets of heads.
+    #[instrument(skip_all, level = tracing::Level::DEBUG)]
+    pub fn get_diff(&self) -> ProjectDiff {
+        tracing::debug!(
+            "branch {:?}, getting changes between {} and {}",
+            self.branch_state.name,
+            self.prev_heads.to_short_form(),
+            self.curr_heads.to_short_form()
+        );
+
+        if self.prev_heads == self.curr_heads {
+            tracing::debug!("no changes");
+            return ProjectDiff::default();
+        }
+
+        let mut diffs: Vec<Diff> = vec![];
+        // Get old and new content
+        let new_file_contents = self.project.get_changed_file_content_between(
+            None,
+            self.branch_state.doc_handle.document_id().clone(),
+            self.prev_heads.clone(),
+            self.curr_heads.clone(),
+            false,
+        );
+        let changed_files_set: HashSet<String> = new_file_contents
+            .iter()
+            .map(|event| match event {
+                FileSystemEvent::FileCreated(path, _) => path.to_string_lossy().to_string(),
+                FileSystemEvent::FileModified(path, _) => path.to_string_lossy().to_string(),
+                FileSystemEvent::FileDeleted(path) => path.to_string_lossy().to_string(),
+            })
+            .collect::<HashSet<String>>();
+        let old_file_contents = self.project.get_files_on_branch_at(
+            self.branch_state,
+            Some(&self.prev_heads),
+            Some(&changed_files_set),
+        );
+
+        for event in &new_file_contents {
+            let (path, new_file_content, change_type) = match event {
```

Comment:
the variant type stuff was an artifact of when we were still including the actual contents in the diff dictionary that we passed back (to know how to convert it), but it's no longer needed; I'd recommend changing new_file_content and old_file_content to `Optional` and just have `FileDeleted` make the content `None`